File: test_regress/t/t_opt_table_sparse_output_split.pl
#!/usr/bin/env perl
if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2003 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

scenarios(simulator => 1);

top_filename("t/t_opt_table_sparse.v");
golden_filename("t/t_opt_table_sparse.out");

compile(
    verilator_flags2 => ["--stats", "--output-split 1"],
    );

if ($Self->{vlt_all}) {
    file_grep($Self->{stats}, qr/Optimizations, Tables created\s+(\d+)/i, 1);
    file_grep($Self->{stats}, qr/ConstPool, Tables emitted\s+(\d+)/i, 2);
}

# Splitting should set VM_PARALLEL_BUILDS to 1 by default
file_grep("$Self->{obj_dir}/$Self->{VM_PREFIX}_classes.mk", qr/VM_PARALLEL_BUILDS\s*=\s*1/);

check_splits(2);

execute(
    check_finished => 1,
    expect_filename => $Self->{golden_filename},
    );

ok(1);
1;

sub check_splits {
    my $expected = shift;
    my $n;
    foreach my $file (glob("$Self->{obj_dir}/*.cpp")) {
        if ($file =~ /__ConstPool_/) {
            $n += 1;
        }
    }
    $n == $expected or error("__ConstPool*.cpp not split: $n");
}

File: test_regress/t/t_opt_table_sparse_output_split.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2024 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('simulator')
test.top_filename = "t/t_opt_table_sparse.v"
test.golden_filename = "t/t_opt_table_sparse.out"


def check_splits(expected):
    n = 0
    for filename in test.glob_some(test.obj_dir + "/*.cpp"):
        if re.search(r'__ConstPool_', filename):
            n += 1
    if n != expected:
        test.error("__ConstPool*.cpp not split: " + str(n))


test.compile(verilator_flags2=["--stats", "--output-split 1"])

if test.vlt_all:
    test.file_grep(test.stats, r'Optimizations, Tables created\s+(\d+)', 1)
    test.file_grep(test.stats, r'ConstPool, Tables emitted\s+(\d+)', 2)

# Splitting should set VM_PARALLEL_BUILDS to 1 by default
test.file_grep(test.obj_dir + "/" + test.vm_prefix + "_classes.mk", r'VM_PARALLEL_BUILDS\s*=\s*1')

check_splits(2)

test.execute(expect_filename=test.golden_filename)

test.passes()
